### PR TITLE
Adds support for the Tartarus Pro (1532:0244)

### LINKED ISSUE #1039: Support for Tartarus Pro (1532:0244)
### Device Info

* Name: Razer Tartarus Pro
* Product Number: RZ07-03110100
* Model Number: RZ07-0311

### Device Information
```
Bus 001 Device 004: ID 1532:0244 Razer USA, Ltd 
Couldn't open device, some information will be missing
Device Descriptor:
  bLength                18
  bDescriptorType         1
  bcdUSB               2.00
  bDeviceClass            0 
  bDeviceSubClass         0 
  bDeviceProtocol         0 
  bMaxPacketSize0        64
  idVendor           0x1532 Razer USA, Ltd
  idProduct          0x0244 
  bcdDevice            2.00
  iManufacturer           1 
  iProduct                2 
  iSerial                 0 
  bNumConfigurations      1
  Configuration Descriptor:
    bLength                 9
    bDescriptorType         2
    wTotalLength       0x0054
    bNumInterfaces          3
    bConfigurationValue     1
    iConfiguration          0 
    bmAttributes         0xa0
      (Bus Powered)
      Remote Wakeup
    MaxPower              500mA
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        0
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      1 Boot Interface Subclass
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      61
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x81  EP 1 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        1
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      1 Keyboard
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength     186
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x82  EP 2 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0018  1x 24 bytes
        bInterval               1
    Interface Descriptor:
      bLength                 9
      bDescriptorType         4
      bInterfaceNumber        2
      bAlternateSetting       0
      bNumEndpoints           1
      bInterfaceClass         3 Human Interface Device
      bInterfaceSubClass      0 
      bInterfaceProtocol      2 Mouse
      iInterface              0 
        HID Device Descriptor:
          bLength                 9
          bDescriptorType        33
          bcdHID               1.11
          bCountryCode            0 Not supported
          bNumDescriptors         1
          bDescriptorType        34 Report
          wDescriptorLength      94
         Report Descriptors: 
           ** UNAVAILABLE **
      Endpoint Descriptor:
        bLength                 7
        bDescriptorType         5
        bEndpointAddress     0x83  EP 3 IN
        bmAttributes            3
          Transfer Type            Interrupt
          Synch Type               None
          Usage Type               Data
        wMaxPacketSize     0x0008  1x 8 bytes
        bInterval               1
```

```
lrwxrwxrwx 1 root root  10 Mar  3 14:54 usb-Razer_Razer_Tartarus_Pro-event-if01 -> ../event28
lrwxrwxrwx 1 root root  10 Mar  3 14:54 usb-Razer_Razer_Tartarus_Pro-event-kbd -> ../event25
lrwxrwxrwx 1 root root  10 Mar  3 14:54 usb-Razer_Razer_Tartarus_Pro-if01-event-kbd -> ../event26
lrwxrwxrwx 1 root root  10 Mar  3 14:54 usb-Razer_Razer_Tartarus_Pro-if02-event-mouse -> ../event30
lrwxrwxrwx 1 root root   9 Mar  3 14:54 usb-Razer_Razer_Tartarus_Pro-if02-mouse -> ../mouse1
```
### Packet Captures

<Please see here as we probably need packet captures>
<https://github.com/openrazer/openrazer/wiki/Reverse-Engineering-USB-Protocol>

### Synapse Screenshots
<Screenshots of all available options as seen on Razer Synapse>
Setup is the same as the Tartarus Chroma, I don't have a screenshot but it is the exact same layout so I suspect that a copy pasta of the Chroma code would be a good place to start. 100% willing to run some tests. will start reading up on the above capture packets.

### PR DESCRIPTION
**Pull request by AlexandreHertzog**
Fixes #1039

Based on the Tartarus V2 codebase. Since many members requested support for this device without the partial key press feature, it was not implemented.

### CODE REVIEW

**Comment by AlexandreHertzog** [comment]
Even though my device is properly recognized, the operations performed at Polychromatic on it take quite a long time and don't really do anything. I'm guessing a timeout happens. Any maintainers willing to review my work and tell me if it is either a coding problem or an environment problem?

File: daemon/openrazer_daemon/hardware/keyboards.py [hunk 1]
```diff
@@ -208,6 +208,50 @@ def _close(self):
...
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+    def _close(self):
+        """
+        Close the key manager
+        """
+        super()._close()
+
+        # self.key_manager.close()
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/keyboards.py]: You can remove both `__init__` and `_close` as they don't do anything here

**Comment by z3ntu** [comment]
Diff looks sane.
Regarding the possible timeout, is there anything in dmesg that would indicate an error from the driver?

**Comment by AlexandreHertzog** [comment]
Thank you very much for the review! I will post the `dmesg` log in a few hours. On the _build_ rules, it is only said that I needed to run the `scripts/build_debs.sh` to generate the install package, in my case. Does this script also compile the C files?

**Comment by AlexandreHertzog** [comment]
Thank you very much for the review! On the _build_ rules, it is only said that I needed to run the `scripts/build_debs.sh` to generate the install package, in my case. Does this script also compile the C files?

Also, immediately when installing my version of openrazer from the generated DEB files and restarting openrazer, my device does not get detected. It is only detected when I fully reboot the system. Is that normal?

Here is the `dmesg` output when trying to interface with the device:

```
[   41.340968] input: key-mapper Razer Razer Tartarus Pro mapped as /devices/virtual/input/input36
[   41.441850] input: key-mapper Razer Razer Tartarus Pro forwarded as /devices/virtual/input/input37
[   41.566978] input: key-mapper Razer Razer Tartarus Pro forwarded as /devices/virtual/input/input38
[   41.668042] input: key-mapper Razer Razer Tartarus Pro forwarded as /devices/virtual/input/input39
[   46.454526] razer driver: Invalid USB response. USB Report length: -110
[   46.454530] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[   51.574514] razer driver: Invalid USB response. USB Report length: -110
[   51.574517] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[   56.694522] razer driver: Invalid USB response. USB Report length: -110
[   56.694526] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[   61.814534] razer driver: Invalid USB response. USB Report length: -110
[   61.814537] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[   66.934536] razer driver: Invalid USB response. USB Report length: -110
[   66.934539] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[   72.054520] razer driver: Invalid USB response. USB Report length: -110
[   72.054523] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[   77.174540] razer driver: Invalid USB response. USB Report length: -110
[   77.174543] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[   82.294554] razer driver: Invalid USB response. USB Report length: -110
[   82.294556] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[   87.418525] razer driver: Invalid USB response. USB Report length: -110
[   87.418529] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[   92.534535] razer driver: Invalid USB response. USB Report length: -110
[   92.534538] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[   97.654571] razer driver: Invalid USB response. USB Report length: -110
[   97.654575] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[  102.774522] razer driver: Invalid USB response. USB Report length: -110
[  102.774526] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[  107.894527] razer driver: Invalid USB response. USB Report length: -110
[  107.894530] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[  114.806546] razer driver: Invalid USB response. USB Report length: -110
[  114.806550] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[  119.930522] razer driver: Invalid USB response. USB Report length: -110
[  119.930527] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[  125.046525] razer driver: Invalid USB response. USB Report length: -110
[  125.046528] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[  130.166546] razer driver: Invalid USB response. USB Report length: -110
[  130.166550] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[  135.286527] razer driver: Invalid USB response. USB Report length: -110
[  135.286530] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
[  135.287330] razer driver: Device data transfer failed.
[  140.406647] razer driver: Invalid USB response. USB Report length: -110
[  140.406651] razerkbd: Invalid Report Length. Start Marker: 00 id: 00 Num Params: 00 Reserved: 00 Command: 00 Params: 00000000000000000000000000000000 .
```

Does that mean that I missed something on my changes?

**Comment by z3ntu** [comment]
Try placing a case statement for your device in `razer_get_report` for the index 0x02

**Comment by AlexandreHertzog** [comment]
I"ll try that, thanks for the suggestion! In the meantime, I migrated from Ubuntu to Fedora. How can I generate the RPM files to test my modifications?

Edit to add: do you have any suggestions as to how can I test my changes without needing to actually reinstall the whole driver?

**Comment by z3ntu** [comment]
Fedora instructions are here https://github.com/openrazer/openrazer/wiki/Building-a-package#fedora

> do you have any suggestions as to how can I test my changes without needing to actually reinstall the whole driver?

You can do `make driver` , `rmmod razerkbd`, `insmod ./driver/razerkbd_driver.ko` and then replug the device so the udev rules fix up the permissions. Then the new driver should be loaded.

**Comment by LunNova** [comment]
Are you likely to keep working on this @AlexandreHertzog ?

I'd like to try out/continue from your starting point if you don't have the time to work on this at the moment.

**Comment by AlexandreHertzog** [comment]
@TransLunarInjection please do! Indeed life got in the way and I couldn't keep working on this. Feel free do fork my work and continue from that if you like.

**Comment by kirill9617** [comment]
I managed to get it working, but with some limitations.
1. In driver mode device sends array of analog values. Now I just translate 06... raw_events to matching 01... . At this point keymapping and threshold level is hardcoded in driver.
2. Profile LEDs are now set together as one RGB led (0xb),  but I have no idea how to read it`s value.

**Comment by shawngmc** [comment]
Is anyone currently working on this? Unfortunately, I don't have an alpine box, but looking at the log it looks like it might just need the fake driver files regenerated, and that the build timed out? I'm not familiar with your build system.

**Comment by z3ntu** [comment]
@shawngmc It literally says at the bottom of the CI output how to fix that error.

But given the code in this PR doesn't seem to work yet, this needs to be fixed first.

**Comment by rahmerh** [comment]
Is there any way to get this working?

**Comment by z3ntu** [comment]
Closing in favor of #1817

**Pull request closed without merging**